### PR TITLE
Image v2

### PR DESCRIPTION
**Pull request by jenswikl**

### CODE REVIEW

**Comment by jenswikl** [comment]
Depends on https://github.com/OP-TEE/build/pull/189

File: bios/main.c [hunk 1]
```diff
@@ -213,52 +193,46 @@ void main_init_sec(struct sec_entry_arg *arg);
...
 	CHECK(hdr.magic != OPTEE_MAGIC || hdr.version != OPTEE_VERSION);
 
 	msg("found secure header\n");
-	sblob_start += sizeof(hdr);
-	CHECK(hdr.init_load_addr_hi != 0);
-
-	pg_part_size = sblob_end - sblob_start - hdr.init_size;
-	pg_part_dst = (size_t)TZ_RES_MEM_START + TZ_RES_MEM_SIZE - pg_part_size;
+	CHECK(hdr.arch != OPTEE_ARCH_ARM32);
+	CHECK(hdr.num_images < 1);
```
Comment (etienne-lms) [inline bios/main.c]:
```c
CHECK(hdr.num_images < 1 ||  hdr.num_images > OPTEE_MAX_NUM_IMAGES);
CHECK(hdr.images[0].image_id != OPTEE_IMAGE_ID_PAGER);
```

File: bios/main.c [hunk 2]
```diff
@@ -168,38 +136,50 @@ static uint32_t copy_dtb(uint32_t dst, uint32_t src)
...
-	dst = copy_bios_image("kernel", kernel_entry, &__linker_nsec_blob_start,
-			&__linker_nsec_blob_end);
+	r = semihosting_download_file("zImage", 64 * 1024 * 1024, kernel_entry);
+	CHECK(r < 0);
+	dst = kernel_entry + r;
 
 	dtb_addr = ROUNDUP(dst, PAGE_SIZE) + 96 * 1024 * 1024; /* safe spot */
 	dst = copy_dtb(dtb_addr, DTB_START);
 
 	rootfs_start = ROUNDUP(dst + DTB_MAX_SIZE, PAGE_SIZE);
```
Comment (etienne-lms) [inline bios/main.c]: minor: (might be out-of-scope of this p-r ?) `dst` is already increased by `DTB_MAX_SIZE`.

File: bios/main.c [hunk 3]
```diff
@@ -213,52 +193,46 @@ void main_init_sec(struct sec_entry_arg *arg);
...
+				      arg->entry);
+	CHECK(r < 0);
 
-	/* Copy secure image in place */
-	copy_bios_image("secure blob", hdr.init_load_addr_lo, sblob_start,
-			sblob_end);
+	if (hdr.num_images > 1) {
+		uint32_t dst = hdr.images[1].load_addr_lo;
+
+		CHECK(hdr.images[1].load_addr_hi != 0xffffffff);
```
Comment (etienne-lms) [inline bios/main.c]: Minor: add `CHECK(hdr.images[1].image_id != OPTEE_IMAGE_ID_PAGED);`

File: bios/main.c [hunk 4]
```diff
@@ -213,52 +193,46 @@ void main_init_sec(struct sec_entry_arg *arg);
...
+		if (dst == 0xffffffff)
+			dst = (size_t)TZ_RES_MEM_START + TZ_RES_MEM_SIZE -
+				hdr.images[1].size;
+		r = semihosting_download_file("tee-pageable_v2.bin",
+					      2 * 1024 * 1024, dst);
```
Comment (etienne-lms) [inline bios/main.c]: minor: increase now the 2MB to something bigger ?
Actually, the max supported size here could be `TZ_RES_MEM_SIZE`.

**Comment by jenswikl** [comment]
Update

File: bios/main.c [hunk 5]
```diff
@@ -213,52 +193,47 @@ void main_init_sec(struct sec_entry_arg *arg);
...
+	CHECK(r < 0);
 
-	/* Copy secure image in place */
-	copy_bios_image("secure blob", hdr.init_load_addr_lo, sblob_start,
-			sblob_end);
+	if (hdr.num_images > 1) {
+		uint32_t dst = hdr.images[1].load_addr_lo;
+
+		CHECK(hdr.images[1].image_id != OPTEE_IMAGE_ID_PAGED);
+		CHECK(hdr.images[1].load_addr_hi != 0xffffffff);
```
Comment (etienne-lms) [inline bios/main.c]: Playing with the format showed me an minor issue: `load_addr_hi` can be null if `dsp != 0xffffffff`.

**Comment by jenswikl** [comment]
Update

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswikl** [comment]
Tag applied

**Pull request merged**